src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -4445,7 +4444,18 @@ FindReferencedTableColumn(Expr *columnExpression, List *parentQueryList, Query *
 	 */
 	if (candidateColumn->varlevelsup > 0)
 	{
-		return;
+		int parentQueryIndex = list_length(parentQueryList) -
+							   candidateColumn->varlevelsup;
+		if (parentQueryIndex < 0)
+		{
+			return;
+		}
+
+		/*
+		 * as we get the query from varlevelsup up, we reset the varlevelsup.
+		 */
+		candidateColumn->varlevelsup = 0;
```

Comment:
are we really allowed to update this var? These are typically the expression from the query tree that we are processing. Shouldn't we need to do:
```C
/*  before we recurse into the query tree, we should update the candidateColumn and we use copy of it. -- followed by your above comment --  */
candidateColumn = copyObject(candidateColumn);
candidateColumn->varlevelsup = 0;

Reply:
You are right, we should copy the variable.